R/tune.R

```diff
@@ -0,0 +1,169 @@
+#' @param metric What metric to use to assess model performance? Options for
+#'   regression: "RMSE" (root-mean-squared error, default), "MAE" (mean-absolute
+#'   error), or "Rsquared." For classification: "ROC" (area under the receiver
+#'   operating characteristic curve).
+#' @param hyperparameters Currently not supported. Optional. A list of
+#'   hyperparameter values to tune over. Overrides \code{tune_depth}. A list of
+#'   lists. The names of the outer-list must match \code{models}. The names of
+#'   each inner-list must match the hyperparameters available to tune over for
+#'   the respective model. Entries in each inner-list are the values of the
+#'   hyperparameter to try. These will be expanded to run a full grid search
+#'   over every combination of values. For details on support models and
+#'   hyperparameters see \code{\link{supported_models}}.
+#'
+#' @return A model_list object
+#' @export
+#'
+#' @importFrom kknn kknn
+#' @importFrom ranger ranger
+#' @importFrom dplyr mutate
+#' @importFrom rlang quo_name
+#'
+#' @details Note that in general a model is trained for each hyperparameter
+#'   combination in each fold for each model, so run time is a function of
+#'   length(models) x n_folds x tune_depth.
+#'
+#' @examples
+#' ### Takes ~10 seconds
+#' \dontrun{
+#' # Remove identifier variable not to be used in model
+#' d <- dplyr::select(pima_diabetes, -PatientID)
+#' # Choose 200 rows to speed tuning
+#' d <- dplyr::sample_n(d, 200)
+#' m <- tune(d, outcome = Diabetes, model_class = "classification")
+#' # Plot performance over hyperparameter values for random forest
+#' ggplot2::ggplot(m$rf)
+#' # Extract confusion matrix for KNN
+#' caret::confusionMatrix(m$knn, norm = "none")
+#' # Compare performance of algorithms at best hyperparameter values
+#' rs <- resamples(m)
+#' dotplot(rs)
+#' }
+tune <- function(d,
+                 outcome,
+                 model_class,
+                 models = c("rf", "knn"),
+                 n_folds = 5,
+                 tune_depth = 10,
+                 tune_method = "random",
+                 metric,
+                 hyperparameters) {
+  # Organize arguments and defaults
+  outcome <- rlang::enquo(outcome)
+  # tibbles upset some algorithms, plus handles matrices, maybe
+  d <- as.data.frame(d)
+  # Make sure outcome's class works with model_class, or infer it
+  outcome_class <- class(dplyr::pull(d, !!outcome))
+  looks_categorical <- outcome_class %in% c("character", "factor")
+  # Some algorithms need the response to be factor instead of char or lgl
+  if (looks_categorical)
+    d <- dplyr::mutate(d, !!rlang::quo_name(outcome) := as.factor(!!outcome))
+  looks_numeric <- is.numeric(dplyr::pull(d, !!outcome))
+  if (!looks_categorical && !looks_numeric) {
+    # outcome is weird class
+    stop(rlang::quo_name(outcome), " is ", class(dplyr::pull(d, !!outcome)),
+         ", and tune doesn't know what to do with that.")
```

Comment:
Convert to numeric or categorical?

Reply:
Did you run into this on an outcome variable that could easily be converted? I thought this would only get hit if outcome was a list or date or other really weird class.